libioc/Provisioning/puppet.py

```diff
@@ -169,91 +171,122 @@ def provision(
         if self.source.remote is True:
-
-            r10kDeployEvent = R10kDeployEvent(
-                scope=_scope,
-                jail=self.jail
-            )
-
-            yield r10kDeployEvent.begin()
-            try:
-                r10k_dir = f"{self.jail.root_path}/usr/local/etc/r10k"
-                r10k_cfg = f"{r10k_dir}/r10k.yaml"
-                if not os.path.isdir(f"{r10k_dir}"):
-                    os.mkdir(r10k_dir, mode=0o755)
-                self.jail.logger.verbose(f"Writing r10k config {r10k_cfg}")
-                with open(r10k_cfg, "w") as f:
-                    f.write("\n".join([f"---",
-                                       ":sources:",
-                                       "    puppet:",
-                                       f"       basedir: {puppet_env_dir}",
-                                       f"       remote: {self.source}\n"]))
-
-                self.jail.logger.verbose("Deploying r10k config")
-                self.jail.exec([
-                    "r10k",
-                    "deploy",
-                    "environment",
-                    "-pv"
-                ])
-            except Exception as e:
-                yield r10kDeployEvent.fail(e)
-                raise e
-            yield r10kDeployEvent.end()
-
-        puppetApplyEvent = PuppetApplyEvent(
-            scope=_scope,
-            jail=self.jail
-        )
-        yield puppetApplyEvent.begin()
-        try:
-            self.jail.exec([
-                "puppet",
-                "apply",
-                "--debug",
-                "--logdest",
-                "syslog",
-                f"{puppet_env_dir}/production/manifests/site.pp"
-            ])
-            yield puppetApplyEvent.end()
-        except Exception as e:
-            yield puppetApplyEvent.fail(e)
-            raise e
-
+            yield from __deploy_r10k(self, event_scope=_scope)
+        yield from __apply_puppet(self, event_scope=_scope)
         if started is True:
-            jailStopEvent = libioc.events.JailStop(
-                jail=self.jail,
-                scope=_scope
-            )
-            yield jailStopEvent.begin()
-            yield from self.jail.stop(event_scope=jailStopEvent.scope)
-            yield jailStopEvent.end()
+            yield from self.jail.stop(event_scope=_scope)
     finally:
         del self.jail.fstab[self.jail.fstab.index(fstab_line)]
 
     yield jailProvisioningEvent.end()
+
+
+def __apply_puppet(
+    self: 'libioc.Provisioning.Prototype',
+    event_scope: libioc.events.Scope
+) -> typing.Generator[R10kDeployEvent, None, None]:
+    puppetApplyEvent = PuppetApplyEvent(
+        scope=event_scope,
+        jail=self.jail
+    )
+    yield puppetApplyEvent.begin()
+    _, _, returncode = self.jail.exec([
+        "puppet",
+        "apply",
+        "--debug",
+        "--logdest",
+        "syslog",
+        f"{__puppet_env_dir}/production/manifests/site.pp"
+    ])
+    if returncode == 0:
```

Comment:
https://puppet.com/docs/puppet/6.4/man/apply.html

we might wanna use `--detailed-exitcodes` here, to better be able to distinguish whether to rerun this, or not